### PR TITLE
Metrics handler

### PR DESCRIPTION
**Pull request by thibaultdvx**
Based on #790.

Refactor of ``MetricsHandler``, based on the changes made on metrics in #790.

``MetricsHandler`` would now behave like this:

``` python
from clinicadl.metrics.handler import MetricsHandler
from clinicadl.metrics.config import AveragePrecisionMetricConfig, LossMetricConfig
from torch.nn import BCELoss
from clinicadl.metrics import Metric

class CustomMetric(Metric): ...

metrics = MetricsHandler(
    loss=LossMetricConfig(loss_fn=BCELoss(reduction="mean")),
    ap=AveragePrecisionMetricConfig(), 
    my_metric=CustomMetric(),
)
```
```python
>>> metrics(batch, epoch=0)
>>> metrics.detailed_df
	epoch	participant	session	sample_id	loss	ap	my_metric
0	0	sub-A	ses-M000	0	0.91	0.42	0.72
1	0	sub-B	ses-M000	0	1.01	0.12	0.81
```
```python
>>> metrics.aggregate(epoch=0)
>>> metrics.df
	epoch	loss	ap	my_metric
0	0	0.9	0.35	0.75
```
In ``df`` you have the metrics aggregated on the whole dataset, whereas in ``detailed_df``, you have the metrics for each element of the dataset.

If it's not in training mode, no need to pass ``epoch``:

```python
>>> metrics(batch)
>>> metrics.detailed_df
	participant	session	sample_id	loss	ap	my_metric
0	sub-A	ses-M000	0	0.91	0.42	0.72
1	sub-B	ses-M000	0	1.01	0.12	0.81
```

```python
>>> metrics.aggregate()
>>> metrics.df
	loss	ap	my_metric
0	0.9	0.35	0.75
```

PS: I had to make some ugly fixes on the Trainer and some callbacks, that I will clean in a next PR.

### CODE REVIEW

File: clinicadl/callbacks/factory/early_stopping.py [hunk 1]
```diff
@@ -167,7 +166,7 @@ def should_stop_training(self, config: _TrainingState, **kwargs) -> bool:
         return False
 
 
-class EarlyStopping(Metrics, Callback):
+class EarlyStopping(Callback):
```
Comment (camillebrianceau) [inline clinicadl/callbacks/factory/early_stopping.py:169]: Don’t we need to add some methods in `EarlyStopping`? I think it inherits from `Metrics` for the `check_metrics` methods

File: clinicadl/callbacks/factory/model_selection.py [hunk 2]
```diff
@@ -5,12 +5,11 @@
 from clinicadl.dictionary.suffixes import PTH, TAR
 from clinicadl.dictionary.words import CHECKPOINT, MODEL, OPTIMIZER
 from clinicadl.metrics.enum import Optimum
-from clinicadl.metrics.handler import Metrics
 
 from .base import Callback
 
 
-class ModelSelection(Callback, Metrics):
+class ModelSelection(Callback):
```
Comment (camillebrianceau) [inline clinicadl/callbacks/factory/model_selection.py:12]: Same here

File: clinicadl/metrics/handler.py [hunk 3]
```diff
@@ -1,249 +1,353 @@
...
-                    metrics_config[metric_name] = LossMetricConfig(
-                        loss_fn=metric.get_object(), reduction=metric.reduction
-                    )
+    metrics: dict[str, MetricOrConfig] = {}
+
+    @field_serializer("metrics")
+    @classmethod
+    def _serialize_metrics(
+        cls, metrics: dict[str, MetricOrConfig]
+    ) -> list[Union[str, dict]]:
```
Comment (camillebrianceau) [inline clinicadl/metrics/handler.py]:
```suggestion
    ) -> dict[str, Union[dict, str]]:
```

**Review by camillebrianceau — APPROVED** [review]
LGTM !

File: clinicadl/callbacks/factory/early_stopping.py (hunk 1 above)
Comment (thibaultdvx) [reply clinicadl/callbacks/factory/early_stopping.py:169]: Ok, I probably missed that. I'll deal with it in a new PR. Thanks
